sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java

```diff
@@ -73,47 +58,38 @@ Instant getOldestMessageTimestamp() {
    * batch is a good bound for future messages.
    */
   @Override
-  public void finalizeCheckpoint() {
-    lock.writeLock().lock();
+  public void finalizeCheckpoint() throws IOException {
     try {
-      for (Message message : messages) {
-        try {
+      LOG.debug("Finalize Checkpoint {} {}", reader, messagesToAck.size());
+      if (reader.active.get() && reader != null) {
+        for (Message message : messagesToAck) {
```

Comment:
Instead of having access to internal reader details, consider exposing this package private method on the reader:
```
public void ackMessages(List<Message> messages) {
...
}
```

This way the internal details of how the watermark advances and thread-safety can be private within the reader.

Reply:
I am going to get back to the initial JmsChekpointmark, it is simple and avoid dependencies between the checkpoint and the reader